x-pack/plugins/security_solution/server/lib/detection_engine/signals/build_rule.ts

```diff
@@ -65,7 +65,7 @@ export const buildRule = ({
 
   const meta = { ...ruleParams.meta, ...riskScoreMeta, ...severityMeta, ...ruleNameMeta };
 
-  const rule = pickBy<RulesSchema>((value: unknown) => value != null, {
```

Comment:
I believe this is the last guard in place to remove any extraneous `null`/`undefined`'s before writing the rule to the signals. In testing everything looks 👌 without (and I believe we've got tests covering this as well), but just want to verify the intent here.

Reply:
Yeah converting the entire rule to `Partial` makes it harder to deal with when building the rest of the signal document since we need `signal.rule.id` later to compute the doc id. My thought is that we should be able to rely on the elasticsearch js library not to convert `undefined` fields to null or anything else so ES should never see `undefined` fields. Explicit `null` fields I think we should handle (usually reject) at the schema level instead of stripping them out later on. Some `null`s are making it into the signal document since they're explicit in the rule SO.

I did more testing on this to figure out where the explicit null fields are coming from and we have some mismatches in the types between signalSchema and RuleTypeParams. For example, timestampOverride is `schema.nullable(schema.string())` (so string | null) in signalSchema, but string | undefined in RuleTypeParams. So what happens when we don't define timestampOverride when creating a rule is it starts as undefined in the request but then the alert plugin validates using signalSchema which coerces the undefined into `null` and stores it in the SO. Then, when we get the rule SO back later to use in the alert executor, timestampOverride is string | null but RuleTypeParams says it's string | undefined and we use it as such. When it's null (which is when we originally sent undefined as part of rule creation), we end up with `signal.rule.timestamp_override: null` in the signal document.

The UI isn't displaying the `null` fields which is probably good. I think the right solution is to resolve the schema mismatches so that our `undefined` values stay undefined instead of being coerced into `null`.

Reply:
Thanks for the extra sleuthing here! 🙂 And yeah, that was my worry with regards to the `null`'s leaking into the signals. ++ to solving this via the schema mismatches to prevent that coercion in the first place.